google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -806,6 +1107,176 @@ public TransactionRunner readWriteTransaction() {
       return new SessionPoolTransactionRunner(SessionPool.this, this);
     }
 
+    @Override
+    public TransactionManager transactionManager() {
+      return new AutoClosingTransactionManager(SessionPool.this, this);
+    }
+
+    @Override
+    public AsyncRunner runAsync() {
+      return new SessionPoolAsyncRunner(SessionPool.this, this);
+    }
+
+    @Override
+    public long executePartitionedUpdate(Statement stmt) {
+      try {
+        return get().executePartitionedUpdate(stmt);
+      } finally {
+        close();
+      }
+    }
+
+    @Override
+    public String getName() {
+      return get().getName();
+    }
+
+    @Override
+    public void prepareReadWriteTransaction() {
+      get().prepareReadWriteTransaction();
+    }
+
+    @Override
+    public void close() {
+      synchronized (lock) {
+        leakedException = null;
+        checkedOutSessions.remove(this);
+      }
+      get().close();
+    }
+
+    @Override
+    public ApiFuture<Empty> asyncClose() {
+      synchronized (lock) {
+        leakedException = null;
+        checkedOutSessions.remove(this);
+      }
+      return get().asyncClose();
+    }
+
+    @Override
+    public PooledSession get() {
+      if (inUse.compareAndSet(false, true)) {
+        try {
+          PooledSession res = super.get();
+          synchronized (lock) {
+            res.markBusy(span);
+            span.addAnnotation(sessionAnnotation(res));
+            incrementNumSessionsInUse();
+            checkedOutSessions.add(this);
+          }
+          initialized.countDown();
+        } catch (Throwable e) {
+          initialized.countDown();
+          // ignore and fallthrough.
```

Comment:
I've restructured the code block a little. The exception is ignored here, because it can only be caused by the call to `super.get()`, and that will be handled in the block below. I've included that in the comment, and restructured the block so that only the call to `super.get()` that can actually cause an exception is in the try-block.